src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -1428,105 +1453,127 @@ private bool TryExecuteSymbolActions(SymbolDeclaredCompilationEvent symbolEvent,
-            CancellationToken cancellationToken,
             out ImmutableArray<DiagnosticAnalyzer> subsetProcessedAnalyzers)
         {
             Debug.Assert(AnalyzerActions.SymbolStartActionsCount > 0);
-            subsetProcessedAnalyzers = default;
 
             var symbol = symbolEvent.Symbol;
             var symbolEndActions = perSymbolActions.SymbolEndActions;
             if (!analysisScope.ShouldAnalyze(symbol) || symbolEndActions.IsEmpty)
             {
                 analysisStateOpt?.MarkSymbolEndAnalysisComplete(symbol, analysisScope.Analyzers);
+                subsetProcessedAnalyzers = ImmutableArray<DiagnosticAnalyzer>.Empty;
                 return true;
             }
 
             var success = true;
-            ArrayBuilder<DiagnosticAnalyzer> subsetProcessedAnalyzersBuilderOpt = null;
-            for (int i = 0; i < analysisScope.Analyzers.Length; i++)
+            var completedAnalyzers = ArrayBuilder<DiagnosticAnalyzer>.GetInstance();
+            var processedAnalyzers = PooledHashSet<DiagnosticAnalyzer>.GetInstance();
+            try
             {
-                var analyzer = analysisScope.Analyzers[i];
-                var analyzerSuccess = true;
-                var symbolEndActionsForAnalyzer = symbolEndActions.Where(a => a.Analyzer == analyzer).ToImmutableArrayOrEmpty();
-                if (!symbolEndActionsForAnalyzer.IsEmpty)
+                foreach (var groupedActions in symbolEndActions.GroupBy(a => a.Analyzer))
                 {
-                    if (!AnalyzerExecutor.TryExecuteSymbolEndActions(symbolEndActionsForAnalyzer, analyzer, symbolEvent, s_GetTopmostNodeForAnalysis, analysisStateOpt))
+                    var analyzer = groupedActions.Key;
+                    if (!analysisScope.Contains(analyzer))
                     {
-                        if (subsetProcessedAnalyzersBuilderOpt == null)
-                        {
-                            subsetProcessedAnalyzersBuilderOpt = ArrayBuilder<DiagnosticAnalyzer>.GetInstance();
-                            subsetProcessedAnalyzersBuilderOpt.AddRange(analysisScope.Analyzers, i);
-                        }
+                        continue;
+                    }
 
-                        analyzerSuccess = false;
+                    processedAnalyzers.Add(analyzer);
+
+                    var symbolEndActionsForAnalyzer = groupedActions.ToImmutableArrayOrEmpty();
+                    if (!symbolEndActionsForAnalyzer.IsEmpty &&
+                        !AnalyzerExecutor.TryExecuteSymbolEndActions(symbolEndActionsForAnalyzer, analyzer, symbolEvent, GetTopmostNodeForAnalysis, analysisStateOpt))
+                    {
                         success = false;
+                        continue;
                     }
-                }
 
-                if (analyzerSuccess)
-                {
                     AnalyzerExecutor.MarkSymbolEndAnalysisComplete(symbol, analyzer, analysisStateOpt);
```

Comment:
>MarkSymbolEndAnalysisComplete [](start = 37, length = 29)

Why call `MarkSymbolEndAnalysisComplete()` for some analyzers here and the rest below? If some need to be executed after this loop, can all be executed after?